test/testsymboldatabase.cpp

```diff
@@ -8786,24 +8786,22 @@ class TestSymbolDatabase : public TestFixture {
         }
         {
             // Container (vector)
-            Settings set;
-            Library::Container vector;
-            vector.startPattern = "Vector <";
-            vector.startPattern2 = "Vector !!::";
```

Comment:
This was wrong. The pattern is actually `Vector < !!::`.

The test now fails at the very first assert in this block so I think we might have a bug on our hands here as the pattern does not seem to make sense as code like `Vector<::` should not appear.

Turns out this was a bug in introduced in an optimization in 16ebb90b3256fc209f3ffe6e79fd4e111d32086e.

Reply:
Actually - the code was wrong before. So it seems it wasn't my change after all.

The test also initially did not include a `<` - that was added by @danmar in f131a99ae30c673e9c304b7e388640a12821f6b1.

The initial code to append this was also added by @danmar in 2a8296879c9f2f43d13b701ee372217cf6c67b8f. The whole changes are unfortunately spread across several commits.

Reply:
That might be a way. But I am not sure if that will fix the test (will try later).

It still seems there is something wrong here as the `startPattern2` will never match in that case.

I assume if `startPattern` ends with `<` we should not generate `startPattern2` at all. But the question is what we should do in that case. Just skip it or drop the trailing ` <`, append `!!::` and match it like the test previously did manually?

The original code before my optimization suggests that something different should have been matched since the `Token::Match()` was redundant which was most likely a copy&paste error.

Reply:
The test strangely works when using `std::vector` which puzzles me. I want to know what the difference is.

Actually. The `!!::` in `startPattern2` would not make any sense since the last token would always match since it would never occur in real code. So in case it ends with a `<` we should not be appending `::!!` at all.

Reply:
I found the error in the changes. I forgot to terminate the encoded `<` character `&lt` - there is the `;` lacking at the end.

Reply:
> Actually. The `!!::` in `startPattern2` would not make any sense since the last token would always match since it would never occur in real code. So in case it ends with a `<` we should not be appending `::!!` at all.

This will fix https://trac.cppcheck.net/ticket/12378
Feel free to take the test case from https://github.com/danmar/cppcheck/pull/5910